### ISSUE TITLE
[$500] Web - Different markdown when copied after sending.

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#26490]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to any report
2. Paste and send markdown provided below
3. Copy the message sent
4. Paste the copied message in composer
5. Observe the difference

Markdown :

![image](https://github.com/Expensify/App/assets/93399543/d644843c-1f63-40d6-9587-3d542a4a04fe)

```
code
```
text 
```
code
```

## Expected Result:
The copied markdown should be the same as the markdown sent.

## Actual Result:
The copied markdown is different from the markdown sent. Specifically, the closing of second code block appears after the text not beneath it:

<img width="138" alt="image" src="https://github.com/Expensify/App/assets/16232057/8f5be27e-c0aa-4fab-a958-1a4b4083b895">

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.3.61.0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/f29cbeb0-1e5d-4625-b1b4-bca9f71d82cc

https://github.com/Expensify/App/assets/93399543/eab6b4e2-11fb-4641-8d72-3374788b9a0b

**Expensify/Expensify Issue URL:**
**Issue reported by:** @krishna2323
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1693066653312359

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0165c30b36785b9c56</li>
        <li>Upwork Job ID: 1711445892210065408</li>
        <li>Last Price Increase: 2023-10-16</li>
    </ul>
</details>

**Comment by trjExpensify** [comment]
I can reproduce this. It produces the same results on enter:

<img width="371" alt="image" src="https://github.com/Expensify/App/assets/16232057/91321d53-375e-42f6-8e76-ed7b4cb8b858">


but yes, the closing of the second code block appears after the text and not beneath it:

<img width="138" alt="image" src="https://github.com/Expensify/App/assets/16232057/d3ddc91d-115e-4848-bc0b-51a38494571d">

I think given that it renders correctly in the chat, it should be classified under [this issue](https://github.com/Expensify/App/issues/20304) and approached in unison with other small markdown fixes. CC: @Beamanator

**Comment by Beamanator** [comment]
That sounds totally fair to me @trjExpensify 👍

**Comment by trjExpensify** [comment]
Done!

**Comment by kbecciv** [comment]
@Krishna2323 Updated

**Comment by trjExpensify** [comment]
Same, @Beamanator is collecting more of these for his master.

**Comment by trjExpensify** [comment]
Alrighty, after a discussion in the linked issue.. we're going to take this off hold. Open for proposals!

**Comment by nikosamofa** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Different markdown when copied after sending. `Enter(\n)` before code block is missing.

### What is the root cause of that problem?

The root cause is in `replacebr` of `expensify-common/ExpensiMark.js/rules` [here](https://github.com/Expensify/expensify-common/blob/main/lib/ExpensiMark.js#L232-L237)

It removes `<br />` just before `<pre>`.

The screenshot shows the trackdown of it.

![need to remove replacebr](https://github.com/Expensify/App/assets/30532000/897ca6c0-bed6-47ee-8d1e-1fe211a25cfc)


After `<br>` added, the new line doesn't show up in html, it's certain that we can remove this rule option.[1]

![br added is fine](https://github.com/Expensify/App/assets/30532000/5a7d3d6c-a85c-4ecd-8c3b-92104d56b7ec)


### What changes do you think we should make in order to solve the problem?

Remove this `replacebr` rule option.

After removing this rule option, the copy/paste works as expected.


https://github.com/Expensify/App/assets/30532000/5c9f5b13-6939-472f-9e00-a6abca8cac7a



I tested on Android chrome and ios safari, it also works as expected, including [1].

**Comment by nikosamofa** [comment]
Contributor details
Your Expensify account email: nikosbrightamofa@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~015f8e8cfd71e80cd1

**Comment by nikosamofa** [comment]
@Beamanator @trjExpensify Can you take a look at the proposal above and give me your feedback?

**Comment by Beamanator** [comment]
@0xmiroslav will be reviewing the proposal 🙏

**Comment by cead22** [comment]
Why not work off this issue?

**Comment by nikosamofa** [comment]
@cead22 created PR https://github.com/Expensify/expensify-common/pull/588
Thanks!

**Comment by nikosamofa** [comment]
@0xmiroslav 
I agree with you that if the old issue won't be fixed, this issue can be ignored as well. If they want to fix it, my solution would be an easier and clearer approach, as it fixes the original html stored in db.

I reviewed the contributing guideline, and so did gpg signing, etc. The PR comment is from expensify-common and I followed the syntax of it, that's why the PR comment is different from other PR comments from App.
If this is not the one you thought, can you kindly explain which part to be updated?

**Comment by cead22** [comment]
> ## Expected Result:
> The copied markdown should be the same as the markdown sent.

I think we've said many times we don't consider this a bug so long as in the end the markdown shown during edit produces the same html as the original, so closing

**Comment by cead22** [comment]
Oh in that case, let's review the proposal and let's fix this! @0xmiroslav can you review @nikosamofa 's proposal please? It looks good to me, though I suggest adding tests to make sure this doesn't come back again

**Comment by nikosamofa** [comment]
@cead22 Thanks and yes, I can add tests for it

**Comment by eh2077** [comment]
Here's an another similar issue shared same root cause https://github.com/Expensify/App/issues/29704

cc @cead22 @0xmiroslav

**Comment by trjExpensify** [comment]
Thanks, @0xmiroslav. How did the proposal review go?

**Comment by nikosamofa** [comment]
@0xmiroslav
cc @trjExpensify 
We have been eagerly awaiting your feedback about the proposals. To keep the momentum going, it would be helpful if you could share your thoughts or any feedback at your earliest convenience.

Thanks

**Comment by trjExpensify** [comment]
@0xmiroslav can you give us an ETA please? It has been [2 weeks](https://github.com/Expensify/App/issues/26490#issuecomment-1771050012) since the last one.

**Comment by nikosamofa** [comment]
> @nikosamofa your solution will cause regression - #8089.

@0xmiroslav, I confirmed it doesn't cause regression for ios/android

https://github.com/Expensify/App/assets/30532000/365ac16c-b621-4ab8-acca-f426a374455c



> Please make all failing automated tests work.
Yes, I can make make all tests pass

**Comment by nikosamofa** [comment]
@0xmiroslav After the fix, the result is the same as your screen recording. So you still want this to be fixed, I'm out

**Comment by trjExpensify** [comment]
https://github.com/Expensify/App/issues/26490#issuecomment-1800118022 - this one shows the issue example solved which is good, so how come it isn't applying to this other case where the three ticks are before and after the text on the same line? (I can see that one on prod as well right now btw).

**Comment by nikosamofa** [comment]
@0xmiroslav It won't fix the issue you mentioned, right
It still fixes https://github.com/Expensify/App/issues/16982

**Comment by cead22** [comment]
@0xmiroslav re: [this comment](https://github.com/Expensify/App/issues/26490#issuecomment-1803845568, I just want to reiterate that it is **not** the case that markdown needs to be the same when copied after sending. As long as the resulting HTML is the same, that's fine

**Comment by nikosamofa** [comment]
Great, I assume the proposal can resolve it. I'm all set to start working once you all agree with it. Thanks

**Comment by eh2077** [comment]
@0xmiroslav Thanks for the ping. Yeah, fixing a bug by deleting code is always tempting!

I just spent some time to restore context about this issue. I think we can't simply delete the `replacebr` rule because it'll introduce regression - adding extra line break on native App. 

You can find more details from my previous [proposal](https://github.com/Expensify/App/issues/16982#issuecomment-1498476588) (I folded the first part of the RCA to make the proposal look shorter).

**Comment by nikosamofa** [comment]
@eh2077 https://github.com/Expensify/App/issues/26490#issuecomment-1800118022
I tried on native App, and it was working fine

**Comment by eh2077** [comment]
> > I think we can't simply delete the `replacebr` rule because it'll introduce regression - adding extra line break on native App.
> 
> I already tested this but not happening on latest codebase. Seems like native issue fixed meanwhile. I suspect #14822 which enabled `enableExperimentalBRCollapsing` on native


Oh, yes, I also can't reproduce it on last main.

**Comment by srikarparsi** [comment]
looks good to me as well thanks @0xmiroslav and @eh2077!

**Comment by eh2077** [comment]
@srikarparsi Yes, this should be assigned to @nikosamofa as mentioned ^

**Comment by trjExpensify** [comment]
I've changed that.

**Comment by trjExpensify** [comment]
I see a expensify-common PR linked, adding the `reviewing` label.

**Comment by nikosamofa** [comment]
@trjExpensify @srikarparsi cc @danieldoglas 
The PR https://github.com/Expensify/expensify-common/pull/606 is merged. What other steps do I need to take to mark this issue as done? Thanks

**Comment by srikarparsi** [comment]
I think once the regression period is over, this issue should be good for payment. Is there anything I'm missing @0xmiroslav?

**Comment by nikosamofa** [comment]
Thank you @0xmiroslav @srikarparsi

**Comment by trjExpensify** [comment]
Yep! Can we get the checklist complete please, then we can start initiate payments. :) 

Confirming those payments when ready will be: 

1. @nikosamofa $500 for the fix
2. @0xmiroslav $500 for the C+ review
3. @Krishna2323 $50 for the bug report

**Comment by nikosamofa** [comment]
@trjExpensify , well, the PR was merged within 2 business days (4 days due to weekend) 😁

**Comment by trjExpensify** [comment]
>@trjExpensify , well, the PR was merged after 2 business days (4 days due to weekend)
>The upwork job says that if it's merged within 3 business days, it's eligible for 50% bonus 😀

[The PR](https://github.com/Expensify/expensify-common/pull/606) was created after Oct 24th, which was the last day of the urgency bonuses as per [this announcement](https://expensify.slack.com/archives/C01GTK53T8Q/p1698191269281379). 

@0xmiroslav sounds good!

**Comment by trjExpensify** [comment]
@Krishna2323 - paid 
@nikosamofa - paid
@0xmiroslav - over to you to track your payment for this one. 

Closing!
